beps/0012-metrics-service/README.md

```diff
@@ -0,0 +1,255 @@
+
+  // Provide access to the underlying meter for advanced configuration.
+  getMeter(): Meter;
+
+  // Additional convenience methods tbd...
+}
+```
+
+The `RootMetricsService` is responsible for the global meter, namespaced to `backstage` or the `serviceName` and `serviceVersion` provided in the app's `app-config.yaml` file. The root metrics service will depend on as little as possible so that it can be initialized as early as possible.
+
+```ts
+export const rootMetricsServiceFactory = createServiceFactory({
+  service: rootMetricsServiceRef,
+  deps: {
+    rootConfig: coreServices.rootConfig,
+  },
+  factory: ({ rootConfig }) => {
+    return DefaultRootMetricsService.fromConfig(rootConfig);
+  },
+});
+```
+
+It also provides a method for creating a plugin-scoped `MetricsService` for each registered plugin.
+
+```ts
+interface RootMetricsService extends MetricsService {
+  forPlugin(pluginId: string): MetricsService;
+}
+
+class DefaultRootMetricsService {
+  // ...
+
+  static fromConfig(config: Config): RootMetricsService {
+    return new DefaultRootMetricsService(config);
+  }
+
+  forPlugin(pluginId: string): MetricsService {
+    return new PluginMetricsService({
+      pluginId,
+      serviceName: this.serviceName,
+      serviceVersion: this.serviceVersion,
+    });
+  }
+}
+```
+
+Each plugin-scoped `MetricsService` acquires a meter when it is created by calling `metrics.getMeter` from the `@opentelemetry/api` package. The internals of the service ensure the metric name is namespaced to match the naming conventions outlined in the [naming conventions](#naming-conventions) section.
+
+```ts
+class PluginMetricsService implements MetricsService {
+  // ...
+  constructor({
+    pluginId,
+    serviceName,
+    serviceVersion,
+  }: PluginMetricsServiceOptions) {
+    this.pluginId = pluginId;
+    this.serviceName = serviceName;
+    this.meter = metrics.getMeter(serviceName, serviceVersion);
+  }
+
+  // Ensures consistency across all metrics
+  private prefixMetricName(name: string): string {
+    return `${this.serviceName}.plugin.${this.pluginId}.${name}`;
+  }
+}
+```
+
+### Naming Conventions
+
+All Backstage metrics MUST follow this hierarchical pattern:
+
+`backstage.{scope}.{scope_name}.{category}.{metric_name}`
+
+**Where:**
+
+- `backstage` is the root namespace for all Backstage metrics
+- `{scope}` is the system scope (`pluginId`, `core` for core services)
+- `{scope_name}` is the name of the scope (e.g. `my-plugin`, `catalog`, `scaffolder`, etc.)
+- `{category}` is the logical grouping within the `{scope}`
+- `{metric_name}` is the name of the metric
+
+The metric name will be the name as provided by the plugin author in the form `{category}.{name}`.
+
+```ts
+const rootMetrics = DefaultRootMetricsService.fromConfig(config);
+const pluginMetricsService = rootMetrics.forPlugin('my-plugin');
+
+const metric = pluginMetricsService.createCounter('my_category.my_metric');
```

Comment:
Can you clarify, not sure if this intended or just rough example code, but when I use `coreServices` in my `deps` like this - https://github.com/backstage/backstage/blob/72f52b6ff964e9f6202874de718d9b8fb0cedc29/plugins/devtools-backend/src/plugin.ts#L33 - I would expect that doing something like `metricsService: coreServices.metricsService` would get me an instance that's coped to my plugin and I would just be off and running making `metricService.createCounter()` calls as needed. Is that the way this will work as well?

Reply:
Example code has been removed.